### ISSUE TITLE
How to use with angularJS

### ISSUE THREAD

**Issue by spanchaluser** [daneWilliams/jquery.dropdown#16]
Can you please let me know how to use this plugin with angularJS?

I have made following directive but the items are not added to the list.

```
app.directive('dropdownjq', function () {
    return {
        restrict: 'A',
        link: function (scope, element, attrs) {
            $(element).dropdown();
        }
    };
});
```

My html binding are as below.

```
<ul class="multi-level" dropdownjq>
    <li data-ng-repeat="item in mylist" data-dropdown-text="item.Text"> </li>
    <li  data-dropdown-text="Item 1"></li>
        <li data-dropdown-text="Item 2">
        <ul>
            <li data-dropdown-text="Item 2.1">
            </li>
            <li data-dropdown-text="Item 2.2">
            </li>
            <li data-dropdown-text="Item 2.3">
            </li>
        </ul>
    </li>
    <li data-dropdown-text="Item 3"></li>
</ul>
```

Item 1, Item 2 and Item 3 are added to the select list but items in my array "mylist" are not added to the select list.

Can anyone please help?

**Comment by daneWilliams** [comment]
I hadn't thought about Angular support to be honest. I'll look into it and maybe have an update for v2. In the meantime, I'll leave this open in case anybody else can help you.

**Comment by zatchgordon** [comment]
This is how I managed to do it

```
app.directive('dropdownjq', dropdownjq)

function dropdownjq($timeout) {
        return {
            restrict: "A",
            require: "ngModel",
            link: link
            }
           function link(scope, element, attrs, ngModel) {
                $(element).hide();
                // add a callback before the timeout so we can apply the scope: see #15 below
                $timeout(function () {
                    var dropdown= $(element).dropdown({options})
                })
          }

}
```

Then In the View

```
<ul class="multi-level" dropdownjq  ng-model="myItem" >
    <li data-ng-repeat="item in mylist" data-dropdown-text="item.Text"> </li>
    <li  data-dropdown-text="Item 1"></li>
        <li data-dropdown-text="Item 2">
        <ul>
            <li data-dropdown-text="Item 2.1">
            </li>
            <li data-dropdown-text="Item 2.2">
            </li>
            <li data-dropdown-text="Item 2.3">
            </li>
        </ul>
    </li>
    <li data-dropdown-text="Item 3"></li>
</ul>
```

The main Difference is:
In the Directive I added a $timeout before adding the jquery and added a require for NgModel.
In the 'ul' tag I add an ng-model (this is required because of the "require: 'ngModel'," in the directive)

For getting information when an item is selected see the link below
[#15](https://github.com/daneWilliams/jquery.dropdown/issues/15)
example 

```
$(element).on( 'dropdown-before-select', function( e, item, previous, dropdown ) {

    // Do stuff here such as:
    // Apply item to the ngModel

});
```

> `item` contains the item object about to be selected, `previous` contains the currently selected item and `dropdown` refers to the current dropdown instance.

After getting a callback on an item change you must then apply that change into angular because jQuery doesn't do this automatically. 

I did this by creating a function called safeApply to make sure I don't interfere with the regular digest cycle

```
                            scope.safeApply = function (fn) {
                                var phase = this.$root.$$phase;
                                if (phase === "$apply" || phase === "$digest") {
                                    if (fn && (typeof (fn) === "function")) {
                                        fn();
                                    }
                                } else {
                                    this.$apply(fn);
                                }
                            };
```

Once an item is obtained from whatever callback function you've chosen it needs to be manually synced with the model.
 assuming the variable "item" is set to the selected item from the callback you can use 

```
                           scope.safeApply(function () {
                                ngModel.$setViewValue(item);
                            }, true);
```

This is how I managed it. It's a bit complicated but this is the only plugin I've found that uses multi-level dropdowns easily.

tl;dr 1) add a timeout
       2) require NgModel
       3) add a callback
       4) manually apply to the NgModel

**Comment by daneWilliams** [comment]
The above methods would need to use the updated event names if using v2+. Please refer to [the Wiki](https://github.com/daneWilliams/jquery.dropdown/wiki/v2-Upgrade) for more information.

**Comment by njcoutinho** [comment]
Is it possible to bind an ng-click on <a> tags in the li elements?

```
<ul  dropdownjq>
    <li  data-dropdown-text="Item 1"><a ng-click="doSomething()">Option A</a></li>
    <li data-dropdown-text="Item 2"><a ng-click="doSomethingElse()">Option B</a></li>
</ul>
```
